themes/src/main/resources/theme/base/admin/resources/templates/kc-provider-config.html

```diff
@@ -27,8 +27,18 @@
             </div>
         </div>
         <div class="col-md-4" data-ng-if="option.type == 'ClientList'">
-            <input type="hidden" ui-select2="clientsUiSelect" id="clients" data-ng-init="initSelectedClient(option.name, config)" data-ng-model="selectedClient" data-ng-change="changeClient(option.name, config, selectedClient);" data-placeholder="{{:: 'selectOne' | translate}}...">
-            </input>
+            <ui-select data-ng-init="initSelectedClient(option.name, config)" data-ng-model="clientUiSelection.selected.newClient"
+                    data-ng-change="changeClient(option.name, config, clientUiSelection.selected.newClient);"
+                    id="clients" theme="bootstrap" style="margin-bottom: 10px;">
+                <ui-select-match placeholder="{{:: 'authz-select-client' | translate}}..." allow-clear="true">{{$select.selected.clientId}}</ui-select-match>
+                <ui-select-choices minimum-input-length="1" refresh="clientsUiSelect($select.search)" refresh-delay="500"
+                        repeat="client as client in clientUiSelection.available.clients">
+                    <div ng-bind-html="client.clientId | highlight: $select.search"/>
+                </ui-select-choices>
+                <ui-select-no-choice>
+                    {{:: 'authz-select-min-character-hint' | translate}}
+                </ui-select-no-choice>
+            </ui-select>
```

Comment:
This is broken for client mappers (present both in clients and IdPs).

Reply:
Will fix it

Reply:
It seems to work for RoleLDAPStorageMapper (which I used some time ago to verify this change)

Reply:
It looks like it is broken in current keycloak version, too :-)
I fixed it for the RoleLDAPStorageMapper where an array is expected by the backend. But the client mappers (e.g. Audience) do expect a single String. Hence: How to proceed? Looks like on of the mappers must be adapted -> I guess the RoleLdap one as it expects an array where only a normal String is making sense from UI perspective.

Reply:
> It looks like it is broken in current keycloak version, too :-)

I didn't investigate it much deeply but I've just checked current master branch and "User Client Role" mapper seem to work fine for me.

Reply:
There are 2 type of users for the JS field type. User federation mappers like the RoleLdap one and client mappers like the Audience one. The RoleLdap mapper uses the JS field to express an array field, the Audience one uses the JS field to model a single String field. This must break and it does. Simply try to create a user federation mapper that uses the client id field. This will result in a failure when saving (example here is from console log):
``
09:04:09,835 ERROR [org.keycloak.services.error.KeycloakErrorHandler] (default task-2) Uncaught server error: com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot deserialize instance of `java.util.ArrayList` out of VALUE_STRING token
 at [Source: (io.undertow.servlet.spec.ServletInputStreamImpl); line: 1, column: 402] (through reference chain: org.keycloak.representations.idm.ComponentRepresentation["config"]->org.keycloak.common.util.MultivaluedHashMap["client.id"])
	at com.fasterxml.jackson.databind.exc.MismatchedInputException.from(MismatchedInputException.java:63)
...
	at org.keycloak.services.filters.KeycloakSessionServletFilter.doFilter(KeycloakSessionServletFilter.java:91)

``

I think it makes sense to fix this by making this field a simple string in all use cases (i.e. adapting the user federation mappers).